2주차 과제 - Counter 앱 & 파일 분리

sejineeee opened the pull request:
## 내용
- Counter 앱 구현
- components 폴더(IncreasingByNumberButton 파일, ClickableButton 파일)
- pages 폴더 (CounterPage 파일)

## 스크린샷


https://user-images.githubusercontent.com/86041335/196435925-0c906927-5a46-441a-be83-81fdd543e33f.mov





## 할 일
- numberList(1, 2, 3, 4, 5 수를 가지고 있는 배열)를 index.jsx 파일에 위치하는게 더 좋을지 아니면 CounterPage에 위치하는게 좋을지 고민해보기

sejineeee committed:
feat: 'Click me' 버튼 컴포넌트 구현

sejineeee committed:
feat: 숫자별 버튼 컴포넌트 구현

sejineeee committed:
feat: Counter page 구현

sejineeee committed:
feat: Counter 앱 구현

sejineeee committed:
feat: 숫자버튼의 수만큼 증가하도록 구현

숫자버튼의 수만큼 증가하도록 구현
Click 버튼을 클릭했을 때 증가하는 숫자를 상수로 선언
메서드에 인자로 값을 전달

daadaadaah commented on src/index.jsx:
```diff
@@ -0,0 +1,26 @@
+
+  return (
+    <div>
+      <CounterPage count={count} handleClick={handleClick} numberList={numberList} />
```
1. props 2개 이상일 경우, 세로정렬해주세용!ㅎㅎ
2. 저는 직접적으로 로직을 처리하는 이벤트 핸들러 함수의 경우 handle~ 을 사용하고 중간에 전달하는 역할 정도만 하는 props 같은 이름으로 사용할 때는 on~ 으로 사용하는 편이에요. 이렇게 하면, 실제 일을 '처리'하는 관점과 '사용'하는 입장에서 바라볼 때의 의도를 각각 드러내줄 수 있거든요!

참고:

- [https://javascript.plainenglish.io/handy-naming-conventions-for-event-handler-functions-props-in-react-fc1cbb791364](https://javascript.plainenglish.io/handy-naming-conventions-for-event-handler-functions-props-in-react-fc1cbb791364)

daadaadaah commented on src/pages/CounterPage.jsx:
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import ClickableButton from '../components/Button/ClickableButton';
```
맥락 구분(라이브러리 import 와 로컬 파일 구분!)을 위해 한줄 띄어줍시당!ㅎㅎ(다른 파일도 마찬가지로요!ㅎㅎ)

daadaadaah commented on src/components/Button/IncreasingByNumberButton.jsx:
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
```
리액트17 버전 이후부터는 생략이 가능해요!
사용하시는 리액트 버전을 확인하신 후 적용해보시면 좋을 것 같아요.
참고:
React 공홈
https://ko.reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#whats-different-in-the-new-transform
기존 JSX 변환과 현재 JSX 변환의 차이점이 잘 나온 링크
https://blog.saeloun.com/2021/07/01/react-17-adds-jsx-runtime-and-jsx-dev-runtime-for-the-new-jsx-transform

daadaadaah commented on src/components/Button/IncreasingByNumberButton.jsx:
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+
+function IncreasingByNumberButton({ number, handleClick }) {
```
`NumberButton`으로만 해도 의도 전달에 문제가 없을것 같아요!ㅎㅎ
컴포넌트 이름 항상 어려운대, 의도만 전달에 문제 없는 수준에서 최소한의 단어를 표현해주는게 좋을 것 같아요!ㅎㅎ

daadaadaah commented on src/index.jsx:
```diff
@@ -0,0 +1,26 @@
+import ReactDOM from 'react-dom';
+import CounterPage from './pages/CounterPage';
+
+function App() {
```
App 컴포넌트는 따로 분리 안해주신 이유가 있으신가요?ㅎㅎ

sejineeee replied:
앗! 따로 분리 안 한 이유는 없습니다!! 수정하여 PR 올리겠습니다!😅

daadaadaah commented on src/components/Button/ClickableButton.jsx:
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+
+function ClickableButton({ count, handleClick }) {
+  const increasingCount = 1;
```
특별히 변수화하신 이유가 있으신가요?

sejineeee replied:
변수화한 이유는 처음에는 1을 적었다가, 이렇게 적어놓으면 어떤 숫자일지 잘 모를 수 있을 거라는 생각이 들어서, 변수화를 하였습니다! 이런 경우에는 변수화를 하지 않는 편이 더 좋은 것인가요? 🥺

daadaadaah replied:
오! 그런 의도시라면, 변수화하는 것도 좋을 것 같아요!ㅎㅎ

daadaadaah reviewed (changes requested):
빠르게 PR 날려주시고, 작업을 작은 단위로 구분해서 commit 메시지로 남긴 점도 너무 잘하셨어요!👏
할일에 아래 사항 적어놓으셨는데, 
> numberList(1, 2, 3, 4, 5 수를 가지고 있는 배열)를 index.jsx 파일에 위치하는게 더 좋을지 아니면 CounterPage에 위치하는게 좋을지 고민해보기

- 비교하실 때, IncreasingByNumberButton 도 같이 비교해보시면 좋을 것 같아요!ㅎㅎ

- 추가로, porps drilling에 대해서 공부해보시는게 좋을 것 같아요!ㅎㅎ

sejineeee committed:
review: 파일명 수정

IncreasingByNumberButton에서 NumberButton으로 파일명 수정

sejineeee committed:
review: 파일명 수정 및 map 메서드 추가

IncreasingByNumberButton에서 NumberButton으로 파일명 수정
numberList 변수 선언 및 배열 할당
map 메서드를 사용하여 button 생성
수정사항에 따라 props 수정

sejineeee committed:
review: 변수에 할당했던 수를 함수에 인자로 전달

변수로 선언해서 할당했던 수를 함수에 인자로 바로 값을 전달
props로 전달받은 props 이름을 handleClick에서 onClick으로 수정

sejineeee committed:
review: useState 추가, 이벤트 핸들러 함수 추가

파일 App에서 관리했던 state와 이벤트 핸들러 함수를 해당 파일로 수정
App에서 전달받았던 props 제거
map 메서드 제거

sejineeee committed:
review: App 컴포넌트 생성

index 파일에 있던 App 컴포넌트를 분리하여 생성

sejineeee committed:
review: App 컴포넌트 분리

sejineeee commented:
### 수정
`Props drilling`에 대해서 찾아보아, `App 컴포넌트`에서 선언했던 state와 이벤트 핸들러 함수를 `CounterPage`에서 선언하여, 각 버튼 컴포넌트로 props로 전달하였습니다.

`numberList` 배열은 `NumberButton` 컴포넌트에서 선언하여 map 메서드를 이용하여 버튼을 구현하도록 수정하였습니다.

`increasingCount`라고 숫자 1을 변수로 선언했던 것은 제거하여, onClick 메서드에 1 값을 전달하는 것으로 수정하였습니다.

### 질문
`import React from 'react'` React 17 이후부터는 import를 할 필요가 없다고 말씀해주셔서, 공식문서 읽어보고 생략해보니, `React is not defined` 에러가 콘솔에 출력됩니다!
공식문서를 보니 `{
  "presets": [
    ["@babel/preset-react", {
      "runtime": "automatic"
    }]
  ]
}` 수정해야 된다고 해서, 수정을 해보았는데도 동일합니다.
(제가 아직 공식문서를 보는 것이 익숙치 않아서, 잘못 봤을 수 있습니다😱)

daadaadaah commented:
> ### 질문
> `import React from 'react'` React 17 이후부터는 import를 할 필요가 없다고 말씀해주셔서, 공식문서 읽어보고 생략해보니, `React is not defined` 에러가 콘솔에 출력됩니다! 공식문서를 보니 `{ "presets": [ ["@babel/preset-react", { "runtime": "automatic" }] ] }` 수정해야 된다고 해서, 수정을 해보았는데도 동일합니다. (제가 아직 공식문서를 보는 것이 익숙치 않아서, 잘못 봤을 수 있습니다😱)

- 엇! 맞는뎅, 한번 수정하신대로 push 해보시겠어요? 제가 clone 해서 한번 테스트 해볼게요!ㅎㅎ

daadaadaah commented on src/App.jsx:
```diff
@@ -0,0 +1,13 @@
+
+import CounterPage from './pages/CounterPage';
+
+function App() {
+  return (
+    <div>
+      <CounterPage />
+    </div>
+  );
+}
```
`<CounterPage />`에 있는 로직들 `<App />` 컴포넌트로 옮김으로써 컴포넌트를 줄일 수 있는데, 따로 `<CounterPage />` 빼주신 이유가 궁금합니다!ㅎㅎ

sejineeee replied:
Props drilling 얘기를 해주셔서, 읽어보았습니다!!
읽어보고, App에서 CounterPage로 Props를 내려주는 것보다는 그러한 측면에서 보았을 때, Counter에서 각 컴포넌트로 내려주는 방식이 더 좋을까 싶어 이렇게 수정해보았습니다!! 🥺

daadaadaah replied:
그렇게 문제를 해결할 수도 있고, CounterPage를 없애는 방법도 있습니다!ㅎㅎ참고하세요!ㅎㅎ

sejineeee committed:
review: babel runtime automatic 추가

sejineeee commented:
> 8f97df9

sejineeee commented:
[8f97df9](https://github.com/CodeSoom/react-week2-assignment-1/pull/180/commits/8f97df96e59af08bf81106fe544e7f0de780ebe4)
babel 파일 수정하여 푸시하였습니다!!

-> 제가 잘못 눌러서 close pull request를 하였습니다. 다시 reopen으로 바꿔놓았는데 따로 문제는 없을까요?

-> (10월 21일 추가 - 추가하고 다시 npm start를 해보니 제대로 동작이 됩니다! 여러 방면으로 확인해보지 않고 질문드려 죄송합니다😭 감사합니다!!)

daadaadaah commented on src/components/Button/ClickableButton.jsx:
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
```
다음 PR에서는 생략해서 문제가 없는지 확인해봅시당!

daadaadaah reviewed (changes requested):
> 제가 잘못 눌러서 close pull request를 하였습니다. 다시 reopen으로 바꿔놓았는데 따로 문제는 없을까요?
- 잘하셨어요!ㅎㅎ이렇게라도 Github 기능 다양하게 써보는거 좋다고 생각합니당!ㅎㅎ

sejineeee committed:
review: counter page 삭제

props drilling 문제로 counter page 삭제

sejineeee committed:
review: react import 생략

sejineeee committed:
review: react import 생략

sejineeee committed:
review: state, handleClick 함수 추가, 각 컴포넌트 호출

counter page 삭제로 인해 state와 handleClick 함수 추가
ClickableButton, NumberButton 컴포넌트 추가 및 props 전달

sejineeee commented:
import react 생략하고, Counter Page를 삭제하고, App 컴포넌트에 작성하는 것으로 수정하였습니다!! 감사합니다~~🙇

daadaadaah reviewed (approved):
수고하셨어요!😊
